Decouple absolute positioned box auto margin distribution from alignment

sammygill opened the pull request:
#### 8d1b743caee7dabba2949d6e40e29b53ca87b19d
<pre>
Decouple absolute positioned box auto margin distribution from alignment
<a href="https://bugs.webkit.org/show_bug.cgi?id=295753">https://bugs.webkit.org/show_bug.cgi?id=295753</a>
<a href="https://rdar.apple.com/155572208">rdar://155572208</a>

Reviewed by Alan Baradlay.

PositionedLayoutConstraints::resolvePosition attempts to resolve the
position of the positioned box. This is done in two main ways:

1.) Distributing the extra space to any auto margins that the box has in
the axis.
2.) Aligning the margin box according to its self-alignment.

Currently, the logic for these two steps is somewhat mixed together. In
particular, if the box has both of its insets fixed, then we will
attempt to see if there are automatic margins to distribute space to and
then attempt to align the box if that fails. Instead, let&apos;s factor this
out so that we can have two distinct sections of code for each of the
steps described above. This is in preparation for building upon the
alignment portion of the code incrementally.

* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
(WebCore::PositionedLayoutConstraints::resolvePosition const):

Canonical link: <a href="https://commits.webkit.org/297267@main">https://commits.webkit.org/297267@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05a9bacdb1a84a4775499bca88648152f7a2d93d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111053 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30719 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21150 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117083 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61320 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113015 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31400 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39301 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84422 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114000 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25083 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99983 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64868 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24427 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18127 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60901 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94463 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18192 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119923 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38102 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28312 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93364 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38478 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96259 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93188 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23759 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38267 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16009 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34083 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37991 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43467 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37655 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40989 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39358 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2eed34aa973fcb8fddfbb370f6f0dcbcd53e8992. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47863

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/05a9bacdb1a84a4775499bca88648152f7a2d93d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47863

sammygill committed:
Decouple absolute positioned box auto margin distribution from alignment
https://bugs.webkit.org/show_bug.cgi?id=295753
rdar://155572208

Reviewed by Alan Baradlay.

PositionedLayoutConstraints::resolvePosition attempts to resolve the
position of the positioned box. This is done in two main ways:

1.) Distributing the extra space to any auto margins that the box has in
the axis.
2.) Aligning the margin box according to its self-alignment.

Currently, the logic for these two steps is somewhat mixed together. In
particular, if the box has both of its insets fixed, then we will
attempt to see if there are automatic margins to distribute space to and
then attempt to align the box if that fails. Instead, let's factor this
out so that we can have two distinct sections of code for each of the
steps described above. This is in preparation for building upon the
alignment portion of the code incrementally.

* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
(WebCore::PositionedLayoutConstraints::resolvePosition const):

Canonical link: https://commits.webkit.org/297267@main
